kive/container/test_dataset_comparison.py

```diff
@@ -0,0 +1,596 @@
+    buffer = io.StringIO(content)
+    return File(name="empty_file.simg", file=buffer)
+
+
+class TestComparisonStrategy(TestCase):
+    "Test the strategy logic in runutils._compare_rerun_datasets"
+
+    def test_apps_must_match(self):
+        original = mock.Mock()
+        rerun = mock.Mock()
+        original.app = 1
+        rerun.app = 2
+        with self.assertRaises(ValueError):
+            runutils.compare_rerun_datasets(original, rerun)
+
+    def test_argtype_must_be_defined(self):
+        arg = mock.Mock()
+        arg.argtype = None
+        run = mock.Mock()
+        run.app.arguments.all = mock.Mock(return_value=[arg])
+        with self.assertRaises(ValueError):
+            runutils.compare_rerun_datasets(run, run)
+
+    def test_argtype_must_be_handled(self):
+        arg = mock.Mock()
+        arg.argtype = "Not none, but not a known value"
+        run = mock.Mock()
+        run.app = mock.Mock()
+        run.app.arguments.all = mock.Mock(return_value=[arg])
+        with self.assertRaises(ValueError):
+            runutils.compare_rerun_datasets(run, run)
+
+    @mock.patch("container.runutils._compare_directory_outputs")
+    @mock.patch("container.runutils._compare_optional_inputs")
+    @mock.patch("container.runutils._compare_monovalent_args")
+    def test_comparison_selection(self, compare_mono, compare_optional,
+                                  compare_dir):
+        "Verify that `compare_rerun_datasets` chooses the right comparison function"
+
+        def reset_mocks():
+            compare_mono.reset_mock()
+            compare_optional.reset_mock()
+            compare_dir.reset_mock()
+
+        def make_mock_run(argtype):
+            arg = mock.Mock()
+            arg.argtype = argtype
+            run = mock.Mock()
+            run.app = mock.Mock()
+            run.app._argument = arg
+            run.app.arguments.all = mock.Mock(return_value=[arg])
+            return run
+
+        cases = [
+            (ContainerArgumentType.FIXED_INPUT, [compare_mono]),
+            (ContainerArgumentType.FIXED_OUTPUT, [compare_mono]),
+            (ContainerArgumentType.OPTIONAL_INPUT, [compare_optional]),
+            (ContainerArgumentType.OPTIONAL_MULTIPLE_INPUT, [compare_optional
+                                                             ]),
+            (ContainerArgumentType.FIXED_DIRECTORY_OUTPUT, [compare_dir]),
+        ]
+
+        for argtype, expect_called in cases:
+            run = make_mock_run(argtype)
+            runutils.compare_rerun_datasets(run, run)
+            for comparison_mock in [
+                    compare_mono, compare_optional, compare_dir
+            ]:
+                if comparison_mock in expect_called:
+                    comparison_mock.assert_called_with(run.app._argument, run, run)
+                else:
+                    comparison_mock.assert_not_called()
+            reset_mocks()
+
+
+class BaseDatasetComparisonTestCase(TestCase):
```

Comment:
I like this way of organizing the test data, and I'm glad to see that `setUpTestData()` was useful. I particularly like your use of `setattr()` and `getattr()` to name the various combinations of datasets and container runs.

Reply:
Thanks 😄 

It does defeat VS Code's static analysis, but I thought it worth it to make the test cases easier to read (and to write).

Reply:
If VS Code is getting cranky, consider putting the datasets into a dictionary instead of attributes on the test class. It takes a bit more typing, but might be worth it. `self.dataset_ab` could become `self.datasets['ab']`.